executor/src/witgen/machines/machine_extractor.rs

```diff
@@ -77,26 +101,45 @@ pub fn split_out_machines<'a, T: FieldElement>(
             .collect::<BTreeMap<_, _>>();
         assert!(connecting_identities.contains_key(&id.id));
 
+        let prover_functions = prover_functions
+            .iter()
+            .copied()
+            .enumerate()
+            .filter(|(_, pf)| {
+                let refs = refs_in_parsed_expression(pf)
+                    .unique()
+                    .filter_map(|n| fixed.column_by_name.get(n).cloned())
+                    .collect::<HashSet<_>>();
+                refs.intersection(&machine_witnesses).next().is_some()
+            })
+            .collect::<Vec<(_, &analyzed::Expression)>>();
+
         log::trace!(
-            "\nExtracted a machine with the following witnesses:\n{} \n and identities:\n{} \n and connecting identities:\n{}",
+            "\nExtracted a machine with the following witnesses:\n{}\n identities:\n{}\n connecting identities:\n{}\n and prover functions:\n{}",
             machine_witnesses
                 .iter()
                 .map(|s| fixed.column_name(s))
                 .sorted()
-                .collect::<Vec<_>>()
-                .join(", "),
+                .format(", "),
             machine_identities
                 .iter()
-                .map(|id| id.to_string())
-                .collect::<Vec<_>>()
-                .join("\n"),
+                .format("\n"),
             connecting_identities
                 .values()
                 .map(|id| id.to_string())
-                .collect::<Vec<_>>()
-                .join("\n"),
+                .format("\n"),
+            prover_functions
+                .iter()
+                .map(|(_, pf)| format!("{pf}"))
+                .format("\n")
         );
 
+        for (i, pf) in &prover_functions {
+            if !extracted_prover_functions.insert(*i) {
+                log::warn!("Prover function was assigned to multiple machines:\n{pf}");
```

Comment:
Does this happen when a prover function references witnesses of two different machines?

Reply:
Yes